runtime/memprof.c

```diff
@@ -654,33 +730,54 @@ void caml_memprof_track_young(uintnat wosize, int from_caml)
   }
 
   /* This condition happens either in the case of an exception or if
-     the callback returned [None]. If these cases happen frequently,
-     then we need to call [flush_deleted] somewhere to prevent a
-     leak. */
-  if (t_idx == Invalid_index)
+     one of the callbacks returned [None]. If these cases happen
+     frequently, then we need to call [flush_deleted] somewhere to
+     prevent a leak. */
+  if (has_delete)
     flush_deleted();
 
-  caml_raise_if_exception(res);
+  if (Is_exception_result(res)) {
```

Comment:
I worry that the exception in `res` might have been collected by `caml_gc_dispatch` at that point. The solution is to extract it and root it before calling the gc. The bug was already there before, I just saw it now.

Reply:
Well spotted, thanks!

@stedolan: we should move the following piece of code:

```C
  /* We can now restore the minor heap in the state needed by
     [Alloc_small_aux]. */
  if (Caml_state->young_ptr - whsize < Caml_state->young_trigger) {
    CAML_INSTR_INT("force_minor/memprof@", 1);
    caml_gc_dispatch();
  }
```

just before `Caml_state->young_ptr -= whsize;` (It is anyway what it is useful for!)

Reply:
Thanks, fixed.